crates/subspace-farmer/src/farming/tests.rs

```diff
@@ -0,0 +1,118 @@
+use crate::commitments::Commitments;
+use crate::farming::Farming;
+use crate::identity::Identity;
+use crate::mock_rpc::MockRpc;
+use crate::plot::Plot;
+use std::sync::Arc;
+use subspace_core_primitives::{Piece, Salt, Tag, TAG_SIZE};
+use subspace_rpc_primitives::SlotInfo;
+use tempfile::TempDir;
+use tokio::sync::mpsc;
+use tokio::time::{sleep, Duration};
+
+fn init() {
+    let _ = env_logger::builder().is_test(true).try_init();
+}
+
+async fn farming_simulator(slots: Vec<SlotInfo>, tags: Vec<Tag>) {
```

Comment:
The way this works is not the same as it works in real life. In real-world use cases you don't push a bunch of stuff right away, you have request/response cycles where new slot challenge is not sent (practically speaking, it can in pathological case) until response for previous slot info is received.

Reply:
These challenges are not sent right away. Slot duration is 1 second at the moment, for the test, I made it 500 millisec (does not matter, can revert it back to 1s). However, to replicate what you suggest, I can accomplish it via:
- when a solution is send, `submit_solution_response` can signal `subscribe_slot_info` via an additional channel
- when `subscribe_slot_info` receives this signal, it can issue the new challenge
Will this structure be okay for mimicking the real-world?

Reply:
That is still not correct. We have challenge/response or request/response here. You send request, wait for response, once response is received you send another challenge. `submit_solution_response` shouldn't signal `subscribe_slot_info`, you should make a request and get a response back, just like Subspace node does it.

Reply:
Ok

Reply:
If I interpreted the workflow correctly, it should be fixed now. Please take a look at the new code to confirm :)